### ISSUE TITLE
Dangerous security bug

### ISSUE THREAD

**Issue by paul1u5** [JohnCoene/firebase#11]
Hi,

just figured out that library has simple but dangerous security bug.

Simply reproduce on DEMO site https://shiny.john-coene.com/firebase/

1. Open https://shiny.john-coene.com/firebase/
2. Go to Chrome browser console (F12) or any other browser
3.  Execute the following lines:

```
        Shiny.setInputValue('fireblaze_' + 'signed_in', {signed_in: true, user: {"email:": "test", emailVerified: true}});
        Shiny.setInputValue('fireblaze_' + 'signed_in_user', {signed_in: true, user: {"email:": "test", emailVerified: true}});
```
4. Thats it. You are  logged in!

Basically user authentication happens on client side in javascript. Then user details are passed to R Shiny server but unfortunately user is not being validated as real user.

To prevent from this see https://firebase.google.com/docs/auth/admin/verify-id-tokens
R shiny backend must validate user token id. It must have valid signature and other details.

Also related material https://en.wikipedia.org/wiki/JSON_Web_Token

**Comment by JohnCoene** [comment]
Oh F*ck!

Thank you for flagging this, I will try to fix this ASAP!

Verify the JWT with R appears to be more painful than I had hoped.

**Comment by JohnCoene** [comment]
I've added a warning to the site and the readme. Sincere apologies!

**Comment by merlinoa** [comment]
This is how we are checking the JWT in R here: https://github.com/Tychobra/polished/blob/bf3bde07c2bfd66f2df37f2c4d4994036e8e57dc/R/Sessions.R#L594

**Comment by JohnCoene** [comment]
Thanks @paul1u5 for reporting this, and thanks @merlinoa for sharing your function, greatly helped me!

I've pushed a fix, will leave it on Github for now waiting for feedback and push to CRAN by EOW.

Sincere apologies again

**Comment by paul1u5** [comment]
@merlinoa function looks quite good but one enhancement I see is:

it is not enough to check first two Google public keys as there might be third keys as well. You can exactly know which key you need as this info is encoded in user token id.

Based on firebase docs, token id has encoded reference id (`kid` value) of public key.

![image](https://user-images.githubusercontent.com/90689948/133431261-26d11c34-ef01-4d3b-990a-4931cbb7c670.png)

JWT token is 
`const token = base64urlEncoding(header) + '.' + base64urlEncoding(payload) + '.' + base64urlEncoding(signature)`

So basically decode header and extract `kid` value. Something like that

```header = base64Decode(str_split(id_token, "\\.")[[1]][1])```

cheers

**Comment by JohnCoene** [comment]
Thanks again for this @paul1u5 I'm going to close this for now. There are warnings on the README and website advising users to update. I'd rather close this so as to limit the visibility of the actual issue.
